File: ppap.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import os
import random
import secrets
import string
import sys

import pyminizip

FILENAME_MAX_SUFFIX = 100
COMPRESS_LEVEL = 9
PASSWORD_MAX_LENGTH = 64
all_chars = string.ascii_letters + string.digits + string.punctuation
PASSWORD_UNSAFE_CHARS = set(" \"'\\/:;*?<>|")
PASSWORD_CHARSET = "".join(c for c in all_chars if c not in PASSWORD_UNSAFE_CHARS)
MESSAGE_TEMPLATE = "generated password is {}"

def create_text_file(text: str, content: str) -> None:
    # Create a text file with the specified content
    with open(text, "w", encoding="utf-8") as file:
        file.write(content)

def encrypt_file(zip: str, password: str, contents: list[str]) -> None:
    # Encrypt a file with a password
    pyminizip.compress(
        contents[0].encode("cp932"), "", zip.encode("cp932"), password, COMPRESS_LEVEL
    )

def generate_password(length: int) -> str:
    # Generate a password of a given length
    if length > PASSWORD_MAX_LENGTH or length < 1:
        length = PASSWORD_MAX_LENGTH
    return "".join(
        secrets.choice(PASSWORD_CHARSET) for i in range(length)
    )

def main(contents: list[str]) -> None:
    # Main function to handle file encryption and password generation
    top: str = contents[0]
    (root, _) = os.path.splitext(top)
    zip: str = root + ".zip"
    txt: str = root + "-password.txt"
    for n in range(FILENAME_MAX_SUFFIX):
        if not os.path.exists(zip) and not os.path.exists(txt):
            break
        zip = f"{root}-{n}.zip"
        txt = f"{root}-{n}-password.txt"
    password: str = generate_password(random.randint(PASSWORD_MAX_LENGTH * 3 // 4, PASSWORD_MAX_LENGTH))
    encrypt_file(zip, password, contents)
    create_text_file(txt, MESSAGE_TEMPLATE.format(password))


if __name__ == "__main__":
    main(sys.argv[1:])

File: ppap.rb
#!/usr/bin/ruby
require 'logger'
require 'optparse'
require 'securerandom'
require 'zip'

FILENAME_MAX_SUFFIX = 100
PASSWORD_MAX_LENGTH = 64
PASSWORD_UNSAFE_CHARS = Set.new(" \"'\\/:;*?<>|".chars)
PASSWORD_CHARSET = ('!'..'~').reject { |c| PASSWORD_UNSAFE_CHARS.include?(c) }

$debug = false
$encryption = true
$logger = nil
$logdev = nil
Zip.default_compression = Zlib::BEST_COMPRESSION
Zip.write_zip64_support = true
# to handle filenames in UTF-8 within the ZIP archive
# Zip.unicode_names = true
FN_ENCODING = 'CP932' # extends Shift_JIS

def logger(materialize = false)
  if $logger.nil?
    $logger = $logdev && (materialize || $debug) ? Logger.new($logdev) : Logger.new(STDOUT)
    $logger.level = $debug ? Logger::DEBUG : Logger::INFO
    $logger.formatter = proc do |severity, datetime, progname, msg|
      "[#{datetime.strftime('%Y-%m-%d %H:%M:%S.%3N')}] [#{severity}] #{msg}\n"
    end
  end
  $logger
end

def setlogger(dev)
  $logdev = dev
  $logger = nil
end

def generate_password(length = 16)
  length = PASSWORD_MAX_LENGTH if length > PASSWORD_MAX_LENGTH || length < 1
  SecureRandom.alphanumeric(length, chars: PASSWORD_CHARSET)
end

def archive(encrypter, archive, files)
  encrypter = Zip::NullEncrypter.new if encrypter.nil?
  Zip::OutputStream.open(archive, encrypter: encrypter) do |zos|
    files.each do |file|
      encoded_name = File.basename(file).encode(FN_ENCODING, invalid: :replace, undef: :replace, replace: '_')
      zos.put_next_entry(encoded_name)
      zos.write(File.read(file, binmode: true))
    end
  end
  logger.debug { "archive created: #{archive}" }
end

def ppap(password, archive, files)
  if archive.to_s.empty?
    extension = File.extname(files[0])
    wo_extension = files[0].delete_suffix(extension)
    archive = "#{wo_extension}.zip"
    if File.exist?(archive)
      (0..FILENAME_MAX_SUFFIX).each do |i|
        archive = "#{wo_extension}-#{i}.zip"
        break unless File.exist?(archive)
      end
    end
    logger.debug { "archive name: #{archive}" }
  end
  setlogger("#{archive.delete_suffix('.zip')}.log")
  encrypter = nil
  if $encryption
    if password.to_s.empty?
      quarter = PASSWORD_MAX_LENGTH / 4
      password = generate_password(SecureRandom.random_number(quarter) + PASSWORD_MAX_LENGTH - quarter + 1)
      logger.debug { "password generated: #{password}" }
    end
    File.open("#{archive.delete_suffix('.zip')}-password.txt", 'w') do |f|
      f.puts("password is #{password}")
      logger.debug { "password file created: #{f.path}" }
    end
    encrypter = Zip::TraditionalEncrypter.new(password)
  end
  archive(encrypter, archive, files)
rescue StandardError => e
  logger(true).error { "error occurred: #{e.message}\n#{e.backtrace.join("\n")}" }
end

if __FILE__ == $0
  archive = nil
  password = nil
  option = OptionParser.new
  option.on('-p', '--password PASSWORD', 'Set password for the zip file') do |p|
    password = p
  end
  option.on('-a', '--archive ARCHIVE', 'Set archive file name') do |a|
    archive = a
  end
  option.on('-d', '--debug', 'Enable debug mode') do
    $debug = true
  end
  option.on('-n', '--no-encryption', 'Disable encryption') do
    $encryption = false
  end
  files = option.parse(ARGV)
  ppap(password, archive, files)
end
